### PR TITLE
Fixing SpatialAveragePooling + tests

### PR DESCRIPTION
**Pull request by fmassa**
There is still the issue with `COUNT_INCLUDE_PAD` that I couldn't figure out.
@szagoruyko, can you have a look at that ?

### CODE REVIEW

**Comment by fmassa** [comment]
Also, @hughperkins, if you could have a look, that's the only thing that is missing for both PRs to be merged.
The issue we have is that it seems that the branch going through `COUNT_INCLUDE_PAD` never considers the `false` case.

**Comment by hughperkins** [comment]
@fmassa, you mean, in both cpu and cuda this issue occurs?  Or just in cuda version?  What does count_include_pad do?

**Comment by hughperkins** [comment]
Seems like count_include_pad means that if some pools go over the edge of the input image, then do we divide by the pooling area inside or outsie the image, is that right?  eg, if the input image is 1x1, and pooling area is 2x2 then:

```
X .
. .
```
- X is a pixel inside image
- both X and . are pixels inside the pooling area

Imagine that X is 3.  Then:
- if count_include_pad: average pool value = 3 / 4 = 0.75
- if not count_include_pad: avearage pool value = 3 / 1 = 3

Is that approximatley right?

**Comment by fmassa** [comment]
@hughperkins the issue only occurs in GPU, CPU works fine.
And the `count_include_pad` is precisely what you described.
I really don't understand why it's not working for GPU, maybe the templated stuff didn't compiled properly in my machine ?

**Comment by hughperkins** [comment]
> the issue only occurs in GPU, CPU works fine.

Ok

> And the count_include_pad is precisely what you described.

Ok

> I really don't understand why it's not working for GPU, maybe the templated stuff didn't compiled properly in my machine ?

Did you try wiping your `~/.nv` ?

**Comment by fmassa** [comment]
Not really. I tried debugging using nvidia-gdb, but I'm not really experienced in GPU programming, so I couldn't go inside the kernel. But what I saw was that the even if `count_include_pad` is `false` and the code calls the template with `false`, it's executed as if it was true.

**Comment by hughperkins** [comment]
Do you have a test that demonstrates that the result is incorrect?

**Comment by hughperkins** [comment]
(just tried running 'nn.testcuda()', and a couple of tests fail from lack of memory, but the no averagepooling tests failed)

**Comment by fmassa** [comment]
Yes, the tests fail for the case `count_include_pad` == 0.
Also, I manually set a toy example which showed that it wasn't working. something like
```lua
m = nn.SpatialAveragePooling(3,3,1,1,1,1):setCountExcludePad():cuda()
m:forward(torch.ones(1,3,3):cuda())
```

**Comment by hughperkins** [comment]
Hmmm, right:
```
SpatialAveragePooling_includepad

 EQ(==) violation   val=0.75, condition=3
	/home/user/torch/install/share/lua/5.1/torch/Tester.lua:42: in function 'asserteq'
	/home/user/torch/install/share/lua/5.1/cunn/test.lua:1731: in function 'v'
```
https://github.com/torch/cunn/commit/fb36141d33c37e3a426eb5ba828ae0984d40b382

**Comment by hughperkins** [comment]
(compiling cunn is very slllloooowwww :-( )

**Comment by fmassa** [comment]
Yes, that's true... but at least it's compiling in parallel.
So, any ideas about why it's not working ?

**Comment by hughperkins** [comment]
- tried removing the `bool` declaration => didnt work
- tried using int instead => didnt work

**Comment by hughperkins** [comment]
Is it because `poolsize` is set to `(hend - hstart) * (wend - wstart)`?  Line 20:
```
    const int pool_size = (hend - hstart) * (wend - wstart);
```
So if we substitute this inside the `if` we get:
```
    if(COUNT_INCLUDE_PAD) {
      top_data[index] = aveval / ((hend - hstart) * (wend - wstart));
    } else {
      top_data[index] = aveval / (hend - hstart) * (wend - wstart);
    }
```

**Comment by fmassa** [comment]
@hughperkins are you using my modifications, which are present in this PR ? I've fixed that

**Comment by hughperkins** [comment]
Ah.  no, I'm using sergey's branch.  Will switch to your branch...

**Comment by hughperkins** [comment]
Seems like hend includes the padding?  Let's say padding is 1, kernel size is 2, stride is 1 and the input image is 1, and no ceil, then we have:

```
o o  .
o X .
.  .  .
```
where:
- X is a pixel in the input image
- o and is the padding area
- o and X is the pooling area

(And lets imagine that pixel at X hsa value 3)

then in the case of no_include_pad, we have:
```
- sum before average = 3
- pool_size = (hend - hstart) * (wend - wstart), where:
  - hend = min(hstart + kernel_h, height + pad_h), where:
     - hstart = ph * stride_h - pad_h, where:
        - ph = index % pooled_width
                 = 0
          pw = (index / pooled_width) % pooled_height
                 = 0
          stride_h = 2
          pad_h = 1
    => hstart = 0 * something - 1 = -1
   hend = min(hstart + kernel_h, height + pad_h)
       = min(-1 + 2, 1 + 1) = min(1, 2) = 1
=> pool_size =  (hend - hstart) * (wend - wstart)
          = (1- (-1)) * (wend-wstart)
         = 2 * 2  (by symmetry, `wend-wstart` = `hend-hstart`)
        = 4
=> avergae = 3 / pool_size = 0.75 = same as include_pad value
```

**Comment by hughperkins** [comment]
(actually, in my example, it is hstart that is in the padding, anyway, basically hstart => hend is kW wide, and includes the padding, I think?)

**Comment by hughperkins** [comment]
eg, easy (but slowish) fix might be maybe something like:

```
const int pool_size = (min(hend, kH -1) - max(hstart,0) * (min(wend, kW -1) - max(wstart,0);
```
?

**Comment by fmassa** [comment]
ohhh, that's true ! the `const int pool_size = (hend - hstart) * (wend - wstart);` comes before the max in `hstart = max(hstart, 0);`. That should indeed solve the problem, just need to put the line where we define pool_size below those max

**Comment by hughperkins** [comment]
Hmmm, right :-)

**Comment by hughperkins** [comment]
Ummm.  surely the ends should be `height - 1` `width - 1`, and not `height`, `width`?

**Comment by hughperkins** [comment]
Sent you a unit test for this.  It failed before; and passes after moving the poolsize down a few rows.

**Comment by fmassa** [comment]
I just pushed a new version in here, could you check that all tests passes ?

**Comment by hughperkins** [comment]
Can I leave you to merge my unit test, and rerun please?  I think it would be good to have a unit test that detects this issue.

**Comment by fmassa** [comment]
@szagoruyko @hughperkins I just realized that for some corner cases, this version (and the `nn` one as well) is not equivalent to Caffe's. I'm preparing a patch for `nn` and `cunn`

**Comment by fmassa** [comment]
@szagoruyko I have tested my fixed version of `nn`s `SpatialAveragePooling`, and I started making necessary changes in the CUDA version as well. I haven't tested this CUDA patch yet, but I think the forward pass should be right. I'm not 100% sure regarding the backward though, will check that later (maybe beginning of next week). If you have the time to try out this version and compare it to my most recent `nn` version, that would be great.

**Comment by fmassa** [comment]
@szagoruyko I think I've fixed the problems with this PR. Tests pass in my machine. Can you have a look at it ?
